File: CS/WindowsFormsApplication1/DocumentLayoutHelper/RichEditDocumentLayoutAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraRichEdit.API.Layout;
using DevExpress.XtraRichEdit.API.Native;

namespace WindowsFormsApplication1.DocumentLayoutHelper {
    public static class RichEditDocumentLayoutAnalyzer {
        public static string GetInformationAboutRichEditDocumentLayout(Document currentDocument, DocumentLayout currentDocumentLayout) {
            SubDocument subDocument = currentDocument.CaretPosition.BeginUpdateDocument();
            DocumentPosition docPosition = subDocument.CreatePosition(currentDocument.CaretPosition.ToInt() == 0 ? 0 : currentDocument.CaretPosition.ToInt() - 1);

            ReadOnlyShapeCollection shapes = subDocument.Shapes.Get(subDocument.CreateRange(docPosition, 1));
            ReadOnlyDocumentImageCollection images = subDocument.Images.Get(subDocument.CreateRange(docPosition, 1));

            if(shapes.Count == 0 && images.Count == 0) docPosition = subDocument.CreatePosition(currentDocument.CaretPosition.ToInt());

            string returnedInformation = "";
            // get infromation about a current document element
            returnedInformation += GetInformationAboutCurrentDocumentElement(currentDocument, currentDocumentLayout, subDocument, docPosition);

            // collect information about CURRENT PAGE
            RangedLayoutElement layoutPosition = currentDocumentLayout.GetElement<RangedLayoutElement>(docPosition);
            if(layoutPosition != null) {
                int currentPageIndex = currentDocumentLayout.GetPageIndex(layoutPosition);
                returnedInformation += PageLayoutHelper.GetInformationAboutCurrentPage(currentDocumentLayout, currentDocumentLayout.GetPage(currentPageIndex), docPosition);                
            }

            currentDocument.CaretPosition.EndUpdateDocument(subDocument);

            return returnedInformation;
        }

        public static string GetInformationAboutCurrentDocumentElement(Document currentDocument, DocumentLayout currentDocumentLayout, SubDocument currentSubDocument, DocumentPosition docPosition) {
            if(currentSubDocument.GetSubDocumentType() == SubDocumentType.TextBox) {
                return TextBoxLayoutHelper.GetInformationAboutCurrentTextBox(currentSubDocument, currentDocumentLayout, docPosition);
            }
            if(currentSubDocument.GetSubDocumentType() == SubDocumentType.Main) {

                RangedLayoutElement tableCell = currentDocumentLayout.GetElement(docPosition, LayoutType.TableCell);
                if(tableCell != null)
                    // collect information about TABLE CELL
                    return DocumentElementLayoutHelper.GetInformationAboutCurrentTableCell(currentDocument, tableCell as LayoutTableCell);

                RangedLayoutElement imageinline = currentDocumentLayout.GetElement(docPosition, LayoutType.InlinePictureBox);
                if(imageinline != null)
                    // collect information about INLINE PICTURE
                    return DocumentElementLayoutHelper.GetInformationAboutCurrentInlinePicture(imageinline as InlinePictureBox);


                RangedLayoutElement floatingObjectAnchor = currentDocumentLayout.GetElement(docPosition, LayoutType.FloatingObjectAnchorBox);
                if(floatingObjectAnchor != null)
                    // collect information about FLOATING OBJECT
                    return DocumentElementLayoutHelper.GetInformationAboutCurrentFloatingObject(floatingObjectAnchor as FloatingObjectAnchorBox, currentDocumentLayout);

                RangedLayoutElement regularTextBlock = currentDocumentLayout.GetElement(docPosition, LayoutType.PlainTextBox);
                if(regularTextBlock != null)
                    // collect information about REGULAR TEXT BLOCK
                    return DocumentElementLayoutHelper.GetInformationAboutRegularTextBlock(regularTextBlock as PlainTextBox, currentDocument);
            }
            return "";
        }
    }
}

File: VB/WindowsFormsApplication1/DocumentLayoutHelper/RichEditDocumentLayoutAnalyzer.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraRichEdit.API.Layout
Imports DevExpress.XtraRichEdit.API.Native

Namespace WindowsFormsApplication1.DocumentLayoutHelper
	Public NotInheritable Class RichEditDocumentLayoutAnalyzer
		Private Sub New()
		End Sub
		Public Shared Function GetInformationAboutRichEditDocumentLayout(ByVal currentDocument As Document, ByVal currentDocumentLayout As DocumentLayout) As String
			Dim subDocument As SubDocument = currentDocument.CaretPosition.BeginUpdateDocument()
			Dim docPosition As DocumentPosition = subDocument.CreatePosition(If(currentDocument.CaretPosition.ToInt() = 0, 0, currentDocument.CaretPosition.ToInt() - 1))

			Dim shapes As ReadOnlyShapeCollection = subDocument.Shapes.Get(subDocument.CreateRange(docPosition, 1))
			Dim images As ReadOnlyDocumentImageCollection = subDocument.Images.Get(subDocument.CreateRange(docPosition, 1))

			If shapes.Count = 0 AndAlso images.Count = 0 Then
				docPosition = subDocument.CreatePosition(currentDocument.CaretPosition.ToInt())
			End If

			Dim returnedInformation As String = ""
			' get infromation about a current document element
			returnedInformation &= GetInformationAboutCurrentDocumentElement(currentDocument, currentDocumentLayout, subDocument, docPosition)

			' collect information about CURRENT PAGE
			Dim layoutPosition As RangedLayoutElement = currentDocumentLayout.GetElement(Of RangedLayoutElement)(docPosition)
			If layoutPosition IsNot Nothing Then
				Dim currentPageIndex As Integer = currentDocumentLayout.GetPageIndex(layoutPosition)
				returnedInformation &= PageLayoutHelper.GetInformationAboutCurrentPage(currentDocumentLayout, currentDocumentLayout.GetPage(currentPageIndex), docPosition)
			End If

			currentDocument.CaretPosition.EndUpdateDocument(subDocument)

			Return returnedInformation
		End Function

		Public Shared Function GetInformationAboutCurrentDocumentElement(ByVal currentDocument As Document, ByVal currentDocumentLayout As DocumentLayout, ByVal currentSubDocument As SubDocument, ByVal docPosition As DocumentPosition) As String
            If currentSubDocument.GetSubDocumentType().Equals(SubDocumentType.TextBox) Then
                Return TextBoxLayoutHelper.GetInformationAboutCurrentTextBox(currentSubDocument, currentDocumentLayout, docPosition)
            End If
			If currentSubDocument.GetSubDocumentType() = SubDocumentType.Main Then

				Dim tableCell As RangedLayoutElement = currentDocumentLayout.GetElement(docPosition, LayoutType.TableCell)
				If tableCell IsNot Nothing Then
					' collect information about TABLE CELL
					Return DocumentElementLayoutHelper.GetInformationAboutCurrentTableCell(currentDocument, TryCast(tableCell, LayoutTableCell))
				End If

				Dim imageinline As RangedLayoutElement = currentDocumentLayout.GetElement(docPosition, LayoutType.InlinePictureBox)
				If imageinline IsNot Nothing Then
					' collect information about INLINE PICTURE
					Return DocumentElementLayoutHelper.GetInformationAboutCurrentInlinePicture(TryCast(imageinline, InlinePictureBox))
				End If


				Dim floatingObjectAnchor As RangedLayoutElement = currentDocumentLayout.GetElement(docPosition, LayoutType.FloatingObjectAnchorBox)
				If floatingObjectAnchor IsNot Nothing Then
                    ' collect information about FLOATING OBJECT
                    Return DocumentElementLayoutHelper.GetInformationAboutCurrentFloatingObject(TryCast(floatingObjectAnchor, FloatingObjectAnchorBox), currentDocumentLayout)
                End If

				Dim regularTextBlock As RangedLayoutElement = currentDocumentLayout.GetElement(docPosition, LayoutType.PlainTextBox)
				If regularTextBlock IsNot Nothing Then
					' collect information about REGULAR TEXT BLOCK
					Return DocumentElementLayoutHelper.GetInformationAboutRegularTextBlock(TryCast(regularTextBlock, PlainTextBox), currentDocument)
				End If
			End If
			Return ""
		End Function
	End Class
End Namespace
